tests/phpunit/PaymentMethods/WC_Stripe_Express_Checkout_Helper_Test.php

```diff
@@ -984,4 +985,79 @@ public function provide_test_has_free_trial() {
 			],
 		];
 	}
+
+	/**
+	 * Test that OPC detection logic works correctly.
+	 *
+	 * @dataProvider provide_opc_detection_scenarios
+	 *
+	 * @param bool  $is_opc Whether the page is detected as One Page Checkout.
+	 * @param array $button_locations Button location settings.
+	 * @param bool  $expected Expected result for should_show_express_checkout_button.
+	 * @group opc
+	 *
+	 * @return void
+	 */
+	public function test_opc_detection_logic( $is_opc, $button_locations, $expected ) {
+		$stripe_settings                                     = WC_Stripe_Helper::get_stripe_settings();
+		$stripe_settings['payment_request_button_locations'] = $button_locations;
+		WC_Stripe_Helper::update_main_stripe_settings( $stripe_settings );
+
+		$gateway = $this->getMockBuilder( WC_Stripe_UPE_Payment_Gateway::class )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$wc_stripe_ece_helper_mock = $this->createPartialMock(
+			WC_Stripe_Express_Checkout_Helper::class,
+			[
+				'is_one_page_checkout',
+				'is_product',
+				'is_checkout',
+				'allowed_items_in_cart',
+				'get_product',
+			],
+			[ $gateway ]
+		);
+
+		// Create a mock product.
+		$product = WC_Helper_Product::create_simple_product();
+		$is_product_page = $is_opc || in_array( 'product', $button_locations, true );
+
+		// Mock the methods.
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_one_page_checkout' )->willReturn( $is_opc );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_product' )->willReturn( $is_product_page );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'is_checkout' )->willReturn( false );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'allowed_items_in_cart' )->willReturn( true );
+		$wc_stripe_ece_helper_mock->expects( $this->any() )->method( 'get_product' )->willReturn( $is_product_page ? $product : false );
+
+		// Manually set the properties that would be set in the constructor.
+		$wc_stripe_ece_helper_mock->stripe_settings = $stripe_settings;
+		$wc_stripe_ece_helper_mock->testmode        = true;
+
+		// Ensure that the 'stripe' gateway is available.
+		$original_gateways                         = WC()->payment_gateways()->payment_gateways;
+		WC()->payment_gateways()->payment_gateways = [
+			'stripe' => new WC_Stripe_UPE_Payment_Gateway(),
+		];
+
+		// Test the actual OPC logic in should_show_express_checkout_button.
+		$result = $wc_stripe_ece_helper_mock->should_show_express_checkout_button();
+
+		$this->assertEquals( $expected, $result );
+
+		// Restore original gateways.
+		WC()->payment_gateways()->payment_gateways = $original_gateways;
+	}
+
+	/**
+	 * Data provider for OPC detection scenarios.
+	 *
+	 * @return array
+	 */
+	public function provide_opc_detection_scenarios() {
+		return [
+			'OPC with checkout enabled'     => [ true, [ 'checkout' ], true ],
+			'Non-OPC with checkout enabled' => [ false, [ 'checkout' ], true ],
```

Comment:
It really feels like we should work out what is going on with the tests so we can add tests for only the `product` page, as well as both the `product` and `checkout` pages.

Reply:
Turns out I updated the wrong setting. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4757/commits/0134bee28ba96da00e755be76eb2e090d8f02d24